### PR TITLE
Add pad op

### PR DESCRIPTION
**Pull request by wanghaoshuang**

### CODE REVIEW

File: paddle/operators/pad_op.h [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    auto* X = context.Input<Tensor>("X");
+    auto* Out = context.Output<Tensor>("Out");
+    Out->mutable_data<T>(context.GetPlace());
+    auto dims = X->dims();
+
+    // Eigen::TensorMap<Eigen::Tensor<const T, 2, Eigen::RowMajor,
+    // Eigen::DenseIndex>> X_tensor = EigenTensor<T, 2>::From(*X);
+    // Eigen::TensorMap<Eigen::Tensor<T, 2, Eigen::RowMajor, Eigen::DenseIndex>>
+    // Out_tensor = EigenTensor<T, 2>::From(*Out);
+    EigenTensor<T, dims.size()>::ConstType X_tensor =
```
Comment (QiJune) [inline paddle/operators/pad_op.h]: 这里的dims的值需要在编译期就确定了，不可以在这里通过传递一个变量来运行时确定。所以PadKernel需要额外多一个模板参数。
```
template<typename Place, typename T, int Dims>
class PadKernel : public framework::OpKernel {
  ...
 EigenTensor<T, Dims>::ConstType x_tensor = EigenTensor<T, Dims>::From(*X);
  ...
};
```

可以参考tensorflow的pad_op的写法

Comment (wanghaoshuang) [reply paddle/operators/pad_op.h]: Thx.

File: paddle/operators/pad_op.h [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License. */
+
+#pragma once
+
+#include "paddle/operators/math/math_function.h"
```
Comment (qingqing01) [inline paddle/operators/pad_op.h]: remove this line.

File: paddle/operators/pad_op.cc [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    for (int i = 0; i < dim0.size(); ++i) {
+      dim1[i] = dim0[i] + paddings[i][0] + paddings[i][1];
+    }
+    ctx.Output<Tensor>("Out")->Resize(dim1);
+  }
+};
+
+class MulOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  MulOpMaker(framework::OpProto *proto, framework::OpAttrChecker *op_checker)
```
Comment (qingqing01) [inline paddle/operators/pad_op.cc]: MulOpMaker -> PadOpMaker

File: paddle/operators/pad_op.cc [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+class PadOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    auto dim0 = ctx.Input<Tensor>("X")->dims();
+    auto dim1 = ctx.Output<Tensor>("Out")->dims();
+    auto paddings = GetAttr<std::vector<std::pair<int32, int32>>>("paddings");
```
Comment (qingqing01) [inline paddle/operators/pad_op.cc]: check the `paddings.size() ==  dim0.size()`

File: python/paddle/v2/framework/tests/test_pad_op.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+                          constant_value=0)
+        }
+
+
+class PadGradOpTest(GradientChecker):
+    def test_pad(self):
+        op = Operator("pad", paddings=((0, 1), (2, 3)), pad_value=0)
+        inputs = {'X': np.random.random((16, 16)).astype("float32"), }
+
+        self.check_grad(op, inputs, set(["X"]), "Out", max_relative_error=0.5)
```
Comment (qingqing01) [inline python/paddle/v2/framework/tests/test_pad_op.py]: add `compare_grad` to compare the CPU and GPU results.

File: paddle/operators/pad_op.h (hunk 1 above)
Comment (qingqing01) [reply paddle/operators/pad_op.h]: 觉得应该是想TensorFlow一样，pad functor带模板Dims(或叫Rank)即可： https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/pad_op.cc#L143

不用 `PadKernel`不用带模板参数Dims

File: paddle/operators/pad_op.h (hunk 2 above)
Comment (wanghaoshuang) [reply paddle/operators/pad_op.h]: Fixed.

File: paddle/operators/pad_op.cc (hunk 3 above)
Comment (wanghaoshuang) [reply paddle/operators/pad_op.cc]: Done.

File: python/paddle/v2/framework/tests/test_pad_op.py (hunk 5 above)
Comment (wanghaoshuang) [reply python/paddle/v2/framework/tests/test_pad_op.py]: Done.

File: paddle/operators/pad_op.cc (hunk 4 above)
Comment (wanghaoshuang) [reply paddle/operators/pad_op.cc]: Done.

File: paddle/operators/pad_op.h [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+using Tensor = framework::Tensor;
+
+template <typename T, size_t D, int MajorType = Eigen::RowMajor,
+          typename IndexType = Eigen::DenseIndex>
+using EigenTensor = framework::EigenTensor<T, D, MajorType, IndexType>;
+
+template <typename Place, typename T, size_t D>
+void PadFunction(const framework::ExecutionContext& context) {
+  auto pads =
+      context.op().GetAttr<std::vector<std::pair<int, int>>>("paddings");
```
Comment (qingqing01) [inline paddle/operators/pad_op.h]: `context.op().` -> `context.`

File: paddle/operators/pad_op.h [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+  Eigen::array<std::pair<int, int>, D> paddings;
+  for (int i = 0; i < pads.size(); ++i) {
+    paddings[i] = pads[i];
+  }
+  T pad_value = context.op().GetAttr<T>("pad_value");
+
+  auto* X = context.Input<Tensor>("X");
```
Comment (qingqing01) [inline paddle/operators/pad_op.h]: The names of variables (including function parameters) and data members are all lowercase

https://google.github.io/styleguide/cppguide.html#Variable_Names

File: paddle/operators/pad_op.h [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+template <typename Place, typename T, size_t D>
+void PadFunction(const framework::ExecutionContext& context) {
+  auto pads =
+      context.op().GetAttr<std::vector<std::pair<int, int>>>("paddings");
+  Eigen::array<std::pair<int, int>, D> paddings;
+  for (int i = 0; i < pads.size(); ++i) {
+    paddings[i] = pads[i];
+  }
+  T pad_value = context.op().GetAttr<T>("pad_value");
```
Comment (qingqing01) [inline paddle/operators/pad_op.h]: context.op() -> context

File: paddle/operators/pad_op.h [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+        PadFunction<Place, T, 4>(context);
+        break;
+      case 5:
+        PadFunction<Place, T, 5>(context);
+        break;
+      case 6:
+        PadFunction<Place, T, 6>(context);
+        break;
+      default:
+        LOG(ERROR) << "Only ranks up to 6 supported.";
```
Comment (qingqing01) [inline paddle/operators/pad_op.h]: LOG(ERROR)  -> PADDLE_THROW(“Only support tensor with ranks up to 6.”)

File: paddle/operators/pad_op.h [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+      case 5:
+        PadGradFunction<Place, T, 5>(context);
+        break;
+      case 6:
+        PadGradFunction<Place, T, 6>(context);
+        break;
+      default:
+        LOG(ERROR) << "Only ranks up to 6 supported.";
+    }
+  }
```
Comment (qingqing01) [inline paddle/operators/pad_op.h]: same problem with PadFunction

File: python/paddle/v2/framework/tests/test_pad_op.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+            type="pad",
+            X="X",
+            Out="Out",
+            paddings=[(0, 1), (2, 3)],
+            pad_value=0)
+        self.inputs = {'X': np.random.random((16, 16)).astype("float32"), }
+
+    def test_normal(self):
+        self.check_grad(
+            self.op, self.inputs, set(["X"]), "Out", max_relative_error=0.5)
```
Comment (qingqing01) [inline python/paddle/v2/framework/tests/test_pad_op.py]: remove `max_relative_error=0.5`. For pad op, the max_relative_error should be very   small.

File: paddle/operators/pad_op.cc [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("X", "The input of pad op");
+    AddOutput("Out", "The output of pad op");
+    AddComment(R"DOC(
+Pad Operator.
+)DOC");
+    AddAttr<std::vector<std::pair<int, int>>>(
+        "paddings", "The padding rules for each dimension");
+    AddAttr<float>("pad_value", "The value to be padded into tensor")
+        .SetDefault(0.0f);
```
Comment (qingqing01) [inline paddle/operators/pad_op.cc]: need more detail and meaningful comments.

https://github.com/PaddlePaddle/Paddle/issues/3885

File: python/paddle/v2/framework/tests/test_pad_op.py [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+        self.type = "pad"
+        self.inputs = {'X': np.random.random((16, 16)).astype("float32"), }
+        self.attrs = {}
+        self.attrs['paddings'] = [(0, 1), (2, 3)]
+        self.attrs['pad_value'] = 0
+        self.outputs = {
+            'Out': np.pad(self.inputs['X'],
+                          self.attrs['paddings'],
+                          mode='constant',
+                          constant_values=0)
```
Comment (qingqing01) [inline python/paddle/v2/framework/tests/test_pad_op.py]: add unit testing for other shape and padding value.

File: paddle/operators/pad_op.h (hunk 6 above)
Comment (wanghaoshuang) [reply paddle/operators/pad_op.h]: Done.

File: paddle/operators/pad_op.h (hunk 8 above)
Comment (wanghaoshuang) [reply paddle/operators/pad_op.h]: Done.

File: paddle/operators/pad_op.h (hunk 7 above)
Comment (wanghaoshuang) [reply paddle/operators/pad_op.h]: Done.

File: paddle/operators/pad_op.h (hunk 9 above)
Comment (wanghaoshuang) [reply paddle/operators/pad_op.h]: Fixed.

File: paddle/operators/pad_op.h (hunk 10 above)
Comment (wanghaoshuang) [reply paddle/operators/pad_op.h]: Fixed.

File: paddle/operators/pad_op.cc (hunk 12 above)
Comment (wanghaoshuang) [reply paddle/operators/pad_op.cc]: Done.

File: paddle/operators/pad_op.cc [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+class PadOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  PadOpMaker(framework::OpProto *proto, framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("X", "The input of pad op");
+    AddOutput("Out", "The output of pad op");
+    AddComment(R"DOC(
+Pad Operator.
+)DOC");
+    AddAttr<std::vector<std::pair<int, int>>>(
```
Comment (QiJune) [inline paddle/operators/pad_op.cc]: 这个attribute是怎么设置成功的呀，我看attribute的定义，目前只能支持

```
typedef boost::variant<boost::blank, int, float, std::string, std::vector<int>,
                       std::vector<float>, std::vector<std::string>>
    Attribute;
```
实际上是没有 std::vector<std::pair<int,int>>的类型的

Comment (wanghaoshuang) [reply paddle/operators/pad_op.cc]: We are discussing about this issue:
https://github.com/PaddlePaddle/Paddle/issues/3877

File: python/paddle/v2/framework/tests/test_pad_op.py (hunk 11 above)
Comment (wanghaoshuang) [reply python/paddle/v2/framework/tests/test_pad_op.py]: Done.

File: python/paddle/v2/framework/tests/test_pad_op.py (hunk 13 above)
Comment (wanghaoshuang) [reply python/paddle/v2/framework/tests/test_pad_op.py]: Done.

File: paddle/operators/pad_op.cc [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+
+using framework::Tensor;
+
+class PadOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    auto dim0 = ctx.Input<Tensor>("X")->dims();
```
[the remark this answers is not in the document]
Comment (wanghaoshuang) [reply paddle/operators/pad_op.cc]: Yes.Thx.

File: paddle/operators/pad_op.cc [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+    auto paddings = GetAttr<std::vector<int>>("paddings");
```
[the remark this answers is not in the document]
Comment (wanghaoshuang) [reply paddle/operators/pad_op.cc]: Fixed. Thx.

File: paddle/operators/pad_op.cc [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+    PADDLE_ENFORCE_EQ(dim0.size(), (int)(paddings.size() / 2),
```
[the remark this answers is not in the document]
Comment (wanghaoshuang) [reply paddle/operators/pad_op.cc]: Thx.

File: paddle/operators/pad_op.cc [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+                      "Size of paddings should be equal to 2 * dimension size "
+                      "of input tensor.");
+    std::vector<int> dim1(dim0.size());
+    for (int i = 0; i < dim0.size(); ++i) {
+      dim1[i] = dim0[i] + paddings[i * 2] + paddings[i * 2 + 1];
```
[the remark this answers is not in the document]
Comment (wanghaoshuang) [reply paddle/operators/pad_op.cc]: Thx.

File: paddle/operators/pad_op.cc [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+    }
+    ctx.Output<Tensor>("Out")->Resize(paddle::framework::make_ddim(dim1));
```
[the remark this answers is not in the document]
Comment (wanghaoshuang) [reply paddle/operators/pad_op.cc]: FIxed.

File: paddle/operators/pad_op.cc [hunk 20]
```diff
@@ -0,0 +1,106 @@
...
+    AddAttr<std::vector<int>>(
+        "paddings",
+        "A list<int> to describes padding rules for each dimension."
+        " For 2-D image tensor, paddings=[0, 1, 2, 3] means"
+        " padding 0 row to top, 1 row to bottom, 2 columns to left"
+        " and 3 columns to right.Size of paddings should be equal to"
+        " 2 * dimension size of input tensor.");
+    AddAttr<float>("pad_value",
+                   "(float) default to 0; "
+                   "The value to be padded into tensor. ")
```
[the remark this answers is not in the document]
Comment (wanghaoshuang) [reply paddle/operators/pad_op.cc]: Fixed.

File: paddle/operators/pad_op.h [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+  for (int i = 0; i < paddings.size(); ++i) {
+    paddings[i].first = pads[i * 2];
+    paddings[i].second = pads[i * 2 + 1];
+  }
+  T pad_value = context.GetAttr<T>("pad_value");
+
+  auto* x = context.Input<Tensor>("X");
+  auto* out = context.Output<Tensor>("Out");
+  out->mutable_data<T>(context.GetPlace());
+  auto dims = x->dims();
```
Comment (Canpio) [inline paddle/operators/pad_op.h]: This `dims` seems unused.

File: paddle/operators/pad_op.h [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+  auto out_tensor = EigenTensor<T, D>::From(*out);
+  auto place = context.GetEigenDevice<Place>();
+  out_tensor.device(place) = x_tensor.pad(paddings, pad_value);
+}
+
+template <typename Place, typename T>
+class PadKernel : public framework::OpKernel {
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    int dim = context.Input<Tensor>("X")->dims().size();
```
Comment (Canpio) [inline paddle/operators/pad_op.h]: The number of a tensor's dim is called `rank` in TensorFlow. We can also use this name.

File: paddle/operators/pad_op.h [hunk 23]
```diff
@@ -0,0 +1,130 @@
...
+        PadFunction<Place, T, 4>(context);
+        break;
+      case 5:
+        PadFunction<Place, T, 5>(context);
+        break;
+      case 6:
+        PadFunction<Place, T, 6>(context);
+        break;
+      default:
+        PADDLE_THROW("Only ranks up to 6 supported.");
```
Comment (Canpio) [inline paddle/operators/pad_op.h]: "PadOp only support tensors with no more than 6 dimensions."
Just for your reference.

File: python/paddle/v2/framework/tests/test_pad_op.py [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+    def initTestCase(self):
+        self.shape = (2, 2, 2)
+        self.paddings = [(0, 0), (0, 0), (1, 2)]
+        self.pad_value = 1
+
+
+class TestCase3(TestPadOp):
+    def initTestCase(self):
+        self.shape = (8)
+        self.paddings = [(0, 1)]
```
Comment (Canpio) [inline python/paddle/v2/framework/tests/test_pad_op.py]: Can `paddings` be negative? If so, we shall add unit tests of it.

File: paddle/operators/pad_op.h (hunk 21 above)
Comment (wanghaoshuang) [reply paddle/operators/pad_op.h]: Fixed.

File: paddle/operators/pad_op.h (hunk 22 above)
Comment (wanghaoshuang) [reply paddle/operators/pad_op.h]: Thks.

File: paddle/operators/pad_op.h (hunk 23 above)
Comment (wanghaoshuang) [reply paddle/operators/pad_op.h]: Thks.

File: paddle/operators/pad_op.cc [hunk 25]
```diff
@@ -0,0 +1,110 @@
...
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("X"), "Input(X) should not be null");
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar(framework::GradVarName("Out")),
+                            "Input(Out@GRAD) should not be null");
+    auto x_dims = ctx.Input<Tensor>("X")->dims();
+    auto *x_grad = ctx.Output<Tensor>(framework::GradVarName("X"));
```
[the remark this answers is not in the document]
Comment (wanghaoshuang) [reply paddle/operators/pad_op.cc]: Fixed.

File: python/paddle/v2/framework/tests/test_pad_op.py (hunk 24 above)
Comment (wanghaoshuang) [reply python/paddle/v2/framework/tests/test_pad_op.py]: Negative paddings is not suggested to be used in `pad op`.  `crop op` is better for your scene.

File: paddle/operators/pad_op.cc [hunk 26]
```diff
@@ -0,0 +1,111 @@
...
+};
+
+class PadOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  PadOpMaker(framework::OpProto *proto, framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("X",
+             "The input of pad op. "
+             "The input should be a k-D tensor(k > 0 and k < 7)");
+    AddOutput("Out",
```
Comment (Canpio) [inline paddle/operators/pad_op.cc]: Seems that output `Out` will not be used in backward infershape and computing? If so, use `NotInGradient()` to mark it, then our graph compiler is able to optimize the memory.

File: paddle/operators/pad_op.cc (hunk 25 above)
[the remark this answers is not in the document]
Comment (Canpio) [reply paddle/operators/pad_op.cc]: Sorry I made a mistake... `x_grad` can be `nullptr`. If so, we don't need to compute `x_grad`.

File: paddle/operators/pad_op.cc (hunk 26 above)
Comment (wanghaoshuang) [reply paddle/operators/pad_op.cc]: FIxed.

**Pull request merged**